datalad_catalog/workflows.py

```diff
@@ -0,0 +1,248 @@
+      - adding translated metadata to a catalog"""
+    # 1. Run dataset-level extraction
+    extracted_file = Path(ds.path) / "extracted_meta.json"
+    for name in extractor_names_dataset:
+        if check_required_files(ds, name):
+            metadata_record = extract_dataset_level(ds, name)
+            write_jsonline_to_file(extracted_file, metadata_record)
+    # 2. Run file-level extraction, add output to same file
+    # - Not implemented yet
+    # 3. Run translation
+    translated_file = Path(ds.path) / "translated_meta.json"
+    with open(extracted_file) as file:
+        for line in file:
+            meta_dict = json.loads(line.rstrip())
+            try:
+                extr_name = meta_dict["extractor_name"]
+                extr_type = meta_dict["type"]
+                mapping_path = schema_dir / get_translation_map(
+                    extr_name, extr_type
+                )
+                write_jsonline_to_file(
+                    translated_file,
+                    translate_to_catalog(meta_dict, mapping_path),
+                )
+            except Exception as e:
+                raise (e)
+    # 4. Add translated metadata to catalog
+    return _add_to_catalog(catalog, translated_file, dict())
+
+
+def extract_dataset_level(dataset, extractor_name):
+    """Extract dataset-level metadata using metalad and a specific extractor"""
+    res = meta_extract(
+        extractorname=extractor_name,
+        dataset=dataset,
+        result_renderer="disabled",
+    )
+    return res[0]["metadata_record"]
+
+
+def conduct_extract_file_level(dataset):
+    """NOT IMPLEMENTED YET"""
+
+
+def extract_file_level(file):
+    """NOT IMPLEMENTED YET"""
+
+
+def check_required_files(dataset: Dataset, extractor_name: str):
+    for fn in required_files[extractor_name]:
+        f_path = Path(dataset.path) / fn
+        if not f_path.exists():
+            return False
+    return True
+
+
+def write_jsonline_to_file(filename, line):
+    """Write a single JSON line to file"""
+    open_as = "a"
+    with open(filename, open_as) as f:
+        json.dump(line, f, cls=jsEncoder)
+        f.write("\n")
+
+
+def translate_to_catalog(meta_obj, mapping_path):
+    """Transforms JSON to JSON using JQ bindings and a translation map
+
+    A translation map should:
+    - be a JSON object written to file
+    - be placed in the 'datalad_catalog/schema/' directory
+    - contain 'variables' and 'properties' keys (respectively, dict and array of dicts)
+
+    Variables are mapped directly from the input object (meta_obj).
+    Properties are mapped from whichever variable is defined as its 'input'.
+    Mapping strings can be constructed using jq syntax: https://stedolan.github.io/jq/
+    """
+    mapping_dict = read_json_file(Path(mapping_path))
+    # Create data variables
+    data_vars = dict()
+    for key in mapping_dict["variables"].keys():
+        data_vars[key] = jq.first(mapping_dict["variables"][key], meta_obj)
```

Comment:
Some `jq` calls will yield an empty result, ending with `StopIteration` (I've seen it with `combinedpersonspubs` on a studyminimeta file without publications - see general comments). My first thought is to catch that:

```suggestion
    for key in mapping_dict["variables"].keys():
        try:
            data_vars[key] = jq.first(mapping_dict["variables"][key], meta_obj)
        except StopIteration:
            data_vars[key] = None
```
But then we need to handle this `None` coming into the next `jq` call below (with `None` as input, some `jq` programs will produce `None`, but some will produce an error). And do we want `None` or an empty string at the end? Does it depend on the context?

Reply:
Repeating my comments in main thread here for clarity:

> Thanks for testing and highlighting all of these. There's an issue for the general case here: https://github.com/datalad/datalad-catalog/issues/151. Indeed the shell workflow handles the null exceptions (and subsequent setting as empty strings, or not) differently from the Python-based workflow-* code. This was likely because the exception handling was difficult to figure out when using the Python jq package (as you stated).
>
> It's worth mentioning here that we shouldn't necessarily be focusing so specifically on jq as a json translator, since it might not be the best package to build the workflow on. Or it could be one of many possible translators. Another could be a specific mapping between the particular fields in the different schemas (i.e. direct mapping). Or using some sort of smart mapping tool. A comment to this extent was made here: https://github.com/datalad/datalad-catalog/issues/91#issuecomment-1189112087. But this has to be explored in detail.
>
> For now, it would perhaps be useful to settle the current practical problem (validation errors) by looking at it from the catalog shema's perspective, identifying what it should be able to process (e.g. "None" in addition to "string" and "array"?), and expecting translators to conform to that. The question is then, should we allow "None" as a rule, or are we just doing that as a workaround because it's difficult to figure out interim exception handling with the jq python package? Both is fine IMO, but better to state that explicitly.